File: src/solutions/2025/03/lua/main.lua
local run = require('src.langs.lua.run')

local options = {
   has_alternate = false,
   has_io = false,
}

---@param input string
---@param is_test boolean
---@param p2 boolean
---@return string
local function solve(input, is_test, p2)
   local res = 0
   for line in input:gmatch('[^\n]+') do
      local le = #line
      local start = 1
      local max = 2
      if (p2) then max = 12 end

      for digit = 1, max do
         local marked = 0
         local n = 0
         local t = max - digit
         local l = le - t
         for i = start, l do
            local parsed = line:sub(i, i):byte() - 48
            if (n < parsed) then
               n = parsed
               marked = i
            end
         end

         start = marked + 1
         res = res + n * (10 ^ t)
      end
   end

   return string.format("%d", res)
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   return solve(input, is_test, false)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   return solve(input, is_test, true)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()

File: src/solutions/2025/03/go/main.go
package main

import (
	core "advent/of/go/src/langs/go"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: false,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

func solve(input string, isTest bool, p2 bool) string {
	res := int64(0)
	for _, line := range strings.Split(input, "\n") {
		start := 0
		max := 2
		if p2 {
			max = 12
		}
		for digit := 0; digit < max; digit++ {
			marked := 0
			n := int64(0)
			t := max - 1 - digit
			l := len(line) - t
			for it := start; it < l; it++ {
				parsed := int64(line[it] - '0')
				if n < parsed {
					marked = it
					n = parsed
				}
			}
			start = marked + 1
			res += n * int64(math.Pow(float64(10), float64(t)))
		}
	}
	return strconv.FormatInt(res, 10)
}

func part1(input string, isTest bool) string {
	return solve(input, isTest, false)
}

func part2(input string, isTest bool) string {
	return solve(input, isTest, true)
}
